### PR TITLE
AZ Charitable Donations

### PR DESCRIPTION
**Pull request by tahsinaislam**

### CODE REVIEW

File: spec/forms/state_file/az_charitable_contributions_form_spec.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    context "with no radio selected" do
+      let(:invalid_params) do
+        {
+          charitable_contributions: "unfilled",
+        }
+      end
+
+      it "returns false" do
+        form = described_class.new(intake, invalid_params)
+        expect(form).not_to be_valid
```
Comment (bengolder) [inline spec/forms/state_file/az_charitable_contributions_form_spec.rb:16]: Non-blocking: when testing validation, I like to add an expectation that an error was added to the specific field I'm trying to test. For example:

```suggestion
      it "returns false" do
        form = described_class.new(intake, invalid_params)
        expect(form).not_to be_valid
        expect(form.errors).to include(:charitable_contributions)
```

This way, if the form is invalid due to a different field, or the error hasn't been properly added to the form's list of errors, then it won't pass.

File: spec/forms/state_file/az_charitable_contributions_form_spec.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        {
+          charitable_contributions: "yes",
+          charitable_cash: 100,
+          charitable_noncash: 600,
+        }
+      end
+
+      it "returns false" do
+        form = described_class.new(intake, invalid_params)
+        expect(form).not_to be_valid
```
Comment (bengolder) [inline spec/forms/state_file/az_charitable_contributions_form_spec.rb:31]:
```suggestion
        expect(form).not_to be_valid
        expect(form.errors).to include(:charitable_noncash)
```

File: spec/forms/state_file/az_charitable_contributions_form_spec.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    it "proceeds with nil amounts for cash and noncash contributions" do
+      form = described_class.new(intake, valid_params)
+      expect(form).to be_valid
+      form.save
+      expect(intake.reload.charitable_cash).to be_nil
+      expect(intake.reload.charitable_noncash).to be_nil
+    end
+  end
+end
```
Comment (bengolder) [inline spec/forms/state_file/az_charitable_contributions_form_spec.rb:94]: Excellent, thanks for adding this test!

File: app/forms/state_file/az_charitable_contributions_form.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    def save
+      if charitable_contributions == "no"
+        @intake.update(charitable_contributions: "no", charitable_cash: nil, charitable_noncash: nil)
+      else
+        @intake.update(attributes_for(:intake))
+      end
+    end
+  end
+end
```
Comment (bengolder) [inline app/forms/state_file/az_charitable_contributions_form.rb:18]: This looks great!

File: app/controllers/state_file/questions/az_charitable_contributions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module StateFile
+  module Questions
+    class AzCharitableContributionsController < QuestionsController
+      private
+
+      def form_params
+        params.require(:state_file_az_charitable_contributions_form).permit([:charitable_contributions, :charitable_cash, :charitable_noncash])
+      end
```
Comment (bengolder) [inline app/controllers/state_file/questions/az_charitable_contributions_controller.rb:8]: I would have thought that these params were permitted automatically by something in the QuestionsController. I believe you can delete this method, unless you tried that before and it failed. If it failed, I'm curious to know why.

**Review by bengolder — APPROVED** [review]

**Review by bengolder — COMMENTED** [review]
Overall, this looks excellent! I had a few small non-blocking questions and suggestions. Great work

File: app/lib/efile/az/az140.rb [hunk 6]
```diff
@@ -34,7 +34,7 @@ def calculate
         set_line(:AZ140_LINE_41, :calculate_line_41)
         set_line(:AZ140_LINE_42, :calculate_line_42)
         set_line(:AZ140_LINE_43, :calculate_line_43)
-        set_line(:AZ140_LINE_43S, -> { "X" }) # TODO if this the way to mark an x on the box?
+        set_line(:AZ140_LINE_43S, :calculate_line_43S) # TODO if this the way to mark an x on the box?
```
Comment (tahsinaislam) [inline app/lib/efile/az/az140.rb:37]: remove todo

**Review by tahsinaislam — COMMENTED** [review]
em approved xml stuff

**Pull request merged**